packages/sdk/react-client/src/client/ClientProvider.tsx

```diff
@@ -72,85 +85,112 @@ export type ClientProviderProps = Pick<ClientContextProps, 'status'> &
-        const config = await getAsyncProviderValue(configProvider);
-        log('resolved config', { config });
-        const services = await getAsyncProviderValue(servicesProvider, config);
-        log('created services', { services });
-        const client = new Client({ config, services, ...options });
-        log('created client');
-        await done(client);
-      }
-    });
+    // Provide external access.
+    useImperativeHandle(forwardedRef, () => client, [client]);
 
-    return () => {
-      log('clean up');
-      clearTimeout(t);
-      void client?.destroy().catch((err) => log.catch(err));
-    };
-  }, [clientProvider, configProvider, servicesProvider]);
+    // Client status subscription.
+    const [status, setStatus] = useControlledValue(controlledStatus);
+    useEffect(() => {
+      if (!client) {
+        return;
+      }
 
-  if (!client || status !== SystemStatus.ACTIVE) {
-    return <Fallback client={client} status={status} />;
-  }
+      const subscription = client.status.subscribe((status) => setStatus(status));
+      return () => subscription.unsubscribe();
+    }, [client]);
+
+    // Create and/or initialize client.
+    useEffect(() => {
+      const initialize = async (client: Client) => {
+        if (!client.initialized) {
+          await client.initialize().catch(setError);
+          log('client ready');
+          types && client.addTypes(types);
+          await onInitialized?.(client);
+          log('initialization complete');
+        } else if (types) {
+          client.addTypes(types);
+        }
+
+        setClient(client);
+
+        // TODO(burdon): Remove since set by hook above?
+        setStatus(client.status.get() ?? SystemStatus.ACTIVE);
+
+        if (!disableBanner) {
+          printBanner(client);
+        }
+      };
+
+      let client: Client;
+      const t = setTimeout(async () => {
+        if (clientProvider) {
+          // Asynchronously request client.
+          client = await getAsyncProviderValue(clientProvider);
+          await initialize(client);
+        } else {
+          // Asynchronously construct client (config may be undefined).
+          const config = await getAsyncProviderValue(configProvider);
+          log('resolved config', { config });
+          const services = await getAsyncProviderValue(servicesProvider, config);
+          log('created services', { services });
+          client = new Client({ config, services, ...options });
+          log('created client');
+          await initialize(client);
+        }
+      });
+
+      return () => {
+        log('clean up');
+        clearTimeout(t);
+        void client
+          ?.destroy()
+          .then(() => {
+            log('destroyed');
+          })
+          .catch((err) => log.catch(err));
+      };
+    }, [configProvider, clientProvider, servicesProvider, disableBanner]);
+
+    if (!client?.initialized || status !== SystemStatus.ACTIVE) {
```

Comment:
if you pass in an initialized `client` and `types` then it will flow through before setting the types.

Reply:
yeah, I included a note about that in the option jsdoc. passing in a client is an advanced thing so I think this is acceptable